Sources/flintc/Compiler.swift

```diff
@@ -16,26 +16,99 @@ import Optimizer
 import IRGen
 
 /// Runs the different stages of the compiler.
-struct Compiler {
-  var inputFiles: [URL]
-  var stdlibFiles: [URL]
-  var outputDirectory: URL
-  var dumpAST: Bool
-  var emitBytecode: Bool
-  var diagnostics: DiagnosticPool
-
-  var sourceContext: SourceContext {
-    return SourceContext(sourceFiles: inputFiles)
+public struct Compiler {
+  private var diagnoseOnly: Bool
+  private var inputFiles: [URL]?
```

Comment:
With the convenience init you don’t have to make all of these nil. To be honest having so many nils is pretty bad. And if there is stuff that might be nil maybe consider splitting up the object, having supporting structs that encapsulate specific compiler modes.